popup/search-results.js

```diff
@@ -0,0 +1,547 @@
+        })
+        .catch(reject);
+    });
+  }
+
+  /**
+   * Fetches style information from userstyles.org's /api/v1/styles/{ID} API.
+   * @param {number} userstylesId The internal "ID" for a style on userstyles.org
+   * @returns {Promise<Object>} An object containing info about the style, e.g. name, author, etc.
+   */
+  function fetchStyle(userstylesId) {
+    return new Promise((resolve, reject) => {
+      const TIMEOUT = 10000;
+      const headers = {
+        'Content-type': 'application/json',
+        'Accept': '*/*'
+      };
+
+      const styleUrl = new URL('https://userstyles.org/api/v1/styles/' + userstylesId);
+      const xhr = new XMLHttpRequest();
+      xhr.timeout = TIMEOUT;
+      xhr.onload = () => {
+        if (xhr.status === 200) {
+          resolve(tryJSONparse(xhr.responseText));
+        } else {
+          console.log('fetch(' + userstylesId + ') [ERROR] ', xhr);
+          reject(xhr.status);
+        }
+      };
+      xhr.onerror = reject;
+      xhr.open('GET', styleUrl, true);
+      for (const key of Object.keys(headers)) {
+        xhr.setRequestHeader(key, headers[key]);
+      }
+      xhr.send();
+    });
+  }
+
+  /**
+   * Fetches (and JSON-parses) search results from a userstyles.org search API.
+   * Automatically sets currentPage and totalPages.
+   * @param {string} category The usrestyles.org "category" (subcategory) OR a any search string.
+   * @return {Object} Response object from userstyles.org
+   */
+  function search(category) {
+    return new Promise((resolve, reject) => {
+      if (totalPages !== undefined && currentPage > totalPages) {
+        resolve({'data':[]});
+      }
+
+      const TIMEOUT = 10000;
+      const headers = {
+        'Content-type': 'application/json',
+        'Accept': '*/*'
+      };
+
+      const searchUrl = new URL('https://userstyles.org/api/v1/styles/subcategory');
+      let queryParams = 'search=' + encodeURIComponent(category);
+      queryParams += '&page=' + currentPage;
+      queryParams += '&country=NA';
+      searchUrl.search = '?' + queryParams;
+      const xhr = new XMLHttpRequest();
+      xhr.timeout = TIMEOUT;
+      xhr.onload = () => {
+        if (xhr.status === 200) {
+          const responseJson = tryJSONparse(xhr.responseText);
+          currentPage = responseJson.current_page + 1;
+          totalPages = responseJson.total_pages;
+          exhausted = (currentPage > totalPages);
+          resolve(responseJson);
+        } else {
+          exhausted = true;
+          reject(xhr.status);
+        }
+      };
+      xhr.onerror = reject;
+      xhr.open('GET', searchUrl, true);
+      for (const key of Object.keys(headers)) {
+        xhr.setRequestHeader(key, headers[key]);
+      }
+      xhr.send();
+    });
+  }
+}
+
+/**
+ * Represents the search results within the Stylus popup.
+ * @returns {Object} Includes load(), next(), and prev() methods to alter the search results.
+ */
+const SearchResults = (() => {
```

Comment:
It's not reusable outside of this feature file so it should be inside one closure above